Librerie/libpdel-0.6.1/pdel/structs/type/structs_type_union.3
.\" @COPYRIGHT@
.\"
.\" Author: Archie Cobbs <archie@freebsd.org>
.\"
.\" $Id: structs_type_union.3 958 2004-12-16 16:49:57Z archie $
.\"
.Dd April 22, 2002
.Dt STRUCTS_TYPE_UNION 3
.Os
.Sh NAME
.Nm structs_type_union
.Nd structs types for C unions
.Sh LIBRARY
PDEL Library (libpdel, \-lpdel)
.Sh SYNOPSIS
.In sys/types.h
.In stddef.h
.In pdel/structs/structs.h
.In pdel/structs/type/union.h
.Fn DEFINE_STRUCTS_UNION struct_name union_name
.Fn STRUCTS_UNION_TYPE union_name field_list
.Fn STRUCTS_UNION_FIELD field_name field_type
.Ft int
.Fn structs_union_set "const struct structs_type *type" "const char *name" "void *data" "const char *field_name"
.Sh DESCRIPTION
The
.Fn STRUCTS_UNION_TYPE
macro defines a
.Xr structs 3
type (i.e., a
.Dv "struct structs_type" )
for describing a
.Li "struct structs_union" ,
which is an intermediate structure describing a
.Li union
.Fa union_name .
.Pp
.Bd -literal -compact -offset 3n
struct structs_union {
    const char  *const field_name; /* name of field currently in use */
    void        *un;               /* pointer to the union itself */
};
.Ed
.Pp
The
.Fa field_name
always indicates which of the union fields is currently in use.
It should never be modified directly; instead, use the function
.Fn structs_union_set
(see below).
.Fa un
points to the actual union, which is stored in a separately allocated buffer
.Xr typed_mem 3
type
.Li "union union_name" .
This buffer is only large enough to hold the specific field currently in use.
.Pp
To define a structure equivalent to a
.Li "struct structs_union"
that declares
.Fa un
to have the correct C type for the union used (instead of
.Li "void *") ,
use the
.Fn DEFINE_STRUCTS_UNION
macro, where
.Fa struct_name
is the name of the structure (or empty if no name is desired) and
.Fa union_name
is the name of the represented union.
Then the
.Fa un
field will be declared to have type
.Fa "union union_name *" .
.Pp
The union's fields are accessible by name.
Of course, only one field may be accessed at a time, and changing the
current union field causes the previous field contents to be lost.
.Pp
As a special case, the field named
.Dq field_name
is also read-only accessible and always returns the name of the union
field currently in use.
The union itself must not contain a field named
.Dq field_name,
or else it will not be accessible.
The
.Dq field_name
field does not appear in the output of
.Xr structs_xml_output 3
or
.Xr structs_traverse 3 .
.Pp
.Fn STRUCTS_UNION_TYPE
defines a
.Xr structs 3
type for a C union.
The
.Fa field_list
parameter must point to an array of
.Li "struct structs_ufield"
structures describing each union field:
.Pp
.Bd -literal -compact -offset 3n
/* This structure describes one field in a union */
struct structs_ufield {
    const char                  *name;       /* field name */
    const struct structs_type   *type;       /* field type */
};
.Ed
.Pp
The
.Fn STRUCTS_UNION_FIELD
macro should be used to define an entry in the field array:
.Fa field_name
is the name of the field and
.Fa field_type
is a pointer to the
.Xr structs 3
type describing the field.
.Pp
The fields need not be listed in the array in the same order as they
are declared in the C union.
However, the array must be terminated with
.Dv STRUCTS_UNION_FIELD_END ,
which is defined as follows:
.Pp
.Dl #define STRUCTS_UNION_FIELD_END { NULL, NULL }
.Pp
The first field in the list is the
.Dq "default field"
and it is chosen as the current field when a union data type is initialized.
When a union data type is read in as XML by
.Xr structs_xml_input 3 ,
if the innermost XML tag is omitted (i.e., the one that specifies the
field name) and the default field has primitive type, then the default
field is assumed.
.Pp
.Fn structs_union_set
should be used to change the union field currently in use,
to guarantee that the old field's contents are properly deallocated
and the new field's contents are properly initialized.
The new field will be initialized to its default value.
The sub-field
.Fa name
of the object pointed to by
.Fa data
must correspond to a
.Li "struct structs_union"
(or equivalent structure), and
.Fa field_name
must match one of the union's field names.
The
.Xr structs_find 3
function may also be used to change the current union field.
.Sh RETURN VALUES
.Fn structs_union_set
returns zero if successful, otherwise -1 with
.Va errno
set appropriately.
.Sh SEE ALSO
.Xr libpdel 3 ,
.Xr structs 3 ,
.Xr structs_type 3 ,
.Xr structs_type_struct 3 ,
.Xr structs_xml_input
.Sh EXAMPLES
The program below sets a union field using the field name and
value specified on the command line:
.Pp
.Bd -literal -compact -offset 3n
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <err.h>

#include <pdel/structs/structs.h>
#include <pdel/structs/type/union.h>
#include <pdel/structs/type/string.h>
#include <pdel/structs/type/ip4.h>
#include <pdel/structs/type/int.h>
#include <pdel/util/typed_mem.h>

/* My union */
union foobar {
        char            *name;
        u_int16_t       index;
        struct in_addr  ipaddr;
};

/* My structs_union structure */
DEFINE_STRUCTS_UNION(foobar_union, foobar);

/* Structs type for a 'struct foobar_union' */
static const struct structs_ufield foobar_fields[] = {
        STRUCTS_UNION_FIELD(name, &structs_type_string),
        STRUCTS_UNION_FIELD(index, &structs_type_uint16),
        STRUCTS_UNION_FIELD(ipaddr, &structs_type_ip4),
        STRUCTS_UNION_FIELD_END
};
static const struct structs_type foobar_type =
        STRUCTS_UNION_TYPE(foobar, &foobar_fields);

static void
show_union(struct foobar_union *un)
{
        /* Show the result */
        if (strcmp(un->field_name, "name") == 0)
                printf("name=\\"%s\\"\\n", un->un->name);
        else if (strcmp(un->field_name, "index") == 0)
                printf("index=%u\\n", un->un->index);
        else if (strcmp(un->field_name, "ipaddr") == 0)
                printf("ipaddr=%s\\n", inet_ntoa(un->un->ipaddr));
        else
                printf("unknown field \\"%s\\"\\n", un->field_name);
}

int
main(int argc, char **argv)
{
        struct foobar_union un;
        const char *name;
        char *value;
        char ebuf[64];

        /* Initialize union */
        if (structs_init(&foobar_type, NULL, &un) == -1)
                err(1, "structs_init");
        printf("Default value: ");
        show_union(&un);

        /* Get the requested field's name and value from command line */
        if (argc != 3)
                errx(1, "usage: setfield <name> <value>");
        name = argv[1];
        value = argv[2];

        /* Set the requested field's value */
        if (structs_set_string(&foobar_type, name,
            value, &un, ebuf, sizeof(ebuf)) == -1)
                errx(1, "%s: %s", name, ebuf);

        /* Show the result */
        printf("New value: ");
        show_union(&un);

        /* Done, clean up */
        structs_free(&foobar_type, NULL, &un);
        return (0);
}
.Ed
.Sh HISTORY
The PDEL library was developed at Packet Design, LLC.
.Dv "http://www.packetdesign.com/"
.Sh AUTHORS
.An Archie Cobbs Aq archie@freebsd.org


Librerie/libpdel-0.6.1/pdel/structs/type/structs_type_union.c

/*
 * @COPYRIGHT@
 *
 * Author: Archie Cobbs <archie@freebsd.org>
 */

#include <sys/types.h>

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <assert.h>
#include <errno.h>

#include "structs/structs.h"
#include "structs/type/array.h"
#include "structs/type/string.h"
#include "structs/type/union.h"
#include "util/typed_mem.h"

/*********************************************************************
			PUBLIC FUNCTIONS
*********************************************************************/

int
structs_union_set(const struct structs_type *type, const char *name,
	void *data, const char *field_name)
{
	const struct structs_ufield *fields;
	const struct structs_ufield *ofield = NULL;
	const struct structs_ufield *field;
	struct structs_union *un;
	const char *mtype;
	void *new_un;
	int have_old = 1;

	/* Find item */
	if ((type = structs_find(type, name, (const void **)&data, 0)) == NULL)
		return (-1);

	/* Sanity check */
	if (type->tclass != STRUCTS_TYPE_UNION) {
		errno = EINVAL;
		return (-1);
	}

	/* Get union info */
	fields = type->args[0].v;
	mtype = type->args[1].s;
	un = data;

	/* Allow 0 initialized union */
	if (un->field_name == NULL &&
	    un->un == NULL) {
	   have_old = 0;
	   goto union_set_new;
	}

	/* Find the old field */
	for (ofield = fields; ofield->name != NULL
	    && strcmp(ofield->name, un->field_name) != 0; ofield++);
	if (ofield->name == NULL) {
		assert(0);
		errno = EINVAL;
		return (-1);
	}

	/* Check if the union is already set to the desired field */
	if (strcmp(un->field_name, field_name) == 0)
		return (0);

 union_set_new:

	/* Find the new field */
	for (field = fields; field->name != NULL
	    && strcmp(field->name, field_name) != 0; field++);
	if (field->name == NULL) {
		errno = ENOENT;
		return (-1);
	}

	/* Create a new union with the new field type */
	if ((new_un = MALLOC(mtype, field->type->size)) == NULL)
		return (-1);
	if ((*field->type->init)(field->type, new_un) == -1) {
		FREE(mtype, new_un);
		return (-1);
	}

	/* Replace existing union with new one having desired type */
	if (have_old) {
	   (*ofield->type->uninit)(ofield->type, un->un);
	   FREE(mtype, un->un);
	}
	un->un = new_un;


	/* Set field name */
	*((const char **)&un->field_name) = field->name;
	return (0);
}

/*********************************************************************
			UNION TYPE METHODS
*********************************************************************/

int
structs_union_init(const struct structs_type *type, void *data)
{
	const struct structs_ufield *const field = type->args[0].v;
	const char *const mtype = type->args[1].s;
	struct structs_union *const un = data;

	/* Sanity */
	if (field->name == NULL) {
		errno = EINVAL;
		return (-1);
	}
	assert(type->tclass == STRUCTS_TYPE_UNION);

	/* Allocate union memory */
	if ((un->un = MALLOC(mtype, field->type->size)) == NULL)
		return (-1);

	/* Initialize field using first member type */
	if ((*field->type->init)(field->type, un->un) == -1) {
		FREE(mtype, un->un);
		return (-1);
	}

	/* Set field name */
	*((const char **)&un->field_name) = field->name;
	return (0);
}

int
structs_union_copy(const struct structs_type *type, const void *from, void *to)
{
	const struct structs_union *const fun = from;
	const struct structs_ufield *const fields = type->args[0].v;
	const struct structs_ufield *field;
	const char *const mtype = type->args[1].s;
	struct structs_union *const tun = to;

	/* Sanity check */
	assert(type->tclass == STRUCTS_TYPE_UNION);

	/* Find field */
	for (field = fields; field->name != NULL
	    && strcmp(fun->field_name, field->name) != 0; field++);
	if (field->name == NULL) {
		assert(0);
		errno = EINVAL;
		return (-1);
	}

	/* Allocate copy union */
	if ((tun->un = MALLOC(mtype, field->type->size)) == NULL)
		return (-1);

	/* Copy field */
	if ((*field->type->copy)(field->type, fun->un, tun->un) == -1) {
		FREE(mtype, tun->un);
		return (-1);
	}

	/* Set field name */
	*((const char **)&tun->field_name) = fun->field_name;
	return (0);
}

int
structs_union_equal(const struct structs_type *type,
	const void *v1, const void *v2)
{
	const struct structs_union *const un1 = v1;
	const struct structs_union *const un2 = v2;
	const struct structs_ufield *const fields = type->args[0].v;
	const struct structs_ufield *field;

	/* Sanity check */
	assert(type->tclass == STRUCTS_TYPE_UNION);

	/* Check if fields are the same */
	if (strcmp(un1->field_name, un2->field_name) != 0)
		return (0);

	/* Find field */
	for (field = fields; field->name != NULL
	    && strcmp(un1->field_name, field->name) != 0; field++);
	if (field->name == NULL) {
		assert(0);
		errno = EINVAL;
		return (-1);
	}

	/* Compare them */
	return ((*field->type->equal)(field->type, un1->un, un2->un));
}

int
structs_union_encode(const struct structs_type *type, const char *mtype,
	struct structs_data *code, const void *data)
{
	const struct structs_union *const un = data;
	const struct structs_ufield *const fields = type->args[0].v;
	const struct structs_ufield *field;
	struct structs_data ncode;
	struct structs_data fcode;

	/* Sanity check */
	assert(type->tclass == STRUCTS_TYPE_UNION);

	/* Find field */
	for (field = fields; field->name != NULL
	    && strcmp(un->field_name, field->name) != 0; field++);
	if (field->name == NULL) {
		assert(0);
		errno = EINVAL;
		return (-1);
	}

	/* Encode name */
	if (structs_string_encode(&structs_type_string,
	    TYPED_MEM_TEMP, &ncode, &field->name) == -1)
		return (-1);

	/* Encode field */
	if ((*field->type->encode)(field->type,
	    TYPED_MEM_TEMP, &fcode, un->un) == -1) {
		FREE(TYPED_MEM_TEMP, ncode.data);
		return (-1);
	}

	/* Allocate code buffer */
	code->length = ncode.length + fcode.length;
	if ((code->data = MALLOC(mtype, code->length)) == NULL) {
		FREE(TYPED_MEM_TEMP, fcode.data);
		FREE(TYPED_MEM_TEMP, ncode.data);
		return (-1);
	}

	/* Copy encoded name and field */
	memcpy(code->data, ncode.data, ncode.length);
	memcpy(code->data + ncode.length, fcode.data, fcode.length);

	/* Done */
	FREE(TYPED_MEM_TEMP, ncode.data);
	FREE(TYPED_MEM_TEMP, fcode.data);
	return (0);
}

int
structs_union_decode(const struct structs_type *type, const u_char *code,
	size_t cmax, void *data, char *ebuf, size_t emax)
{
	const struct structs_ufield *const fields = type->args[0].v;
	const char *const mtype = type->args[1].s;
	struct structs_union *const un = data;
	const struct structs_ufield *field;
	char *field_name;
	int nlen;
	int flen;

	/* Sanity check */
	assert(type->tclass == STRUCTS_TYPE_UNION);

	/* Decode field name */
	if ((nlen = structs_string_decode(&structs_type_string,
	    code, cmax, &field_name, ebuf, emax)) == -1)
		return (-1);

	/* Find field */
	for (field = fields; field->name != NULL
	    && strcmp(field_name, field->name) != 0; field++);
	if (field->name == NULL) {
		snprintf(ebuf, emax, "unknown union field \"%s\"", field_name);
		FREE(STRUCTS_TYPE_STRING_MTYPE, field_name);
		return (-1);
	}
	FREE(STRUCTS_TYPE_STRING_MTYPE, field_name);

	/* Allocate field memory */
	if ((un->un = MALLOC(mtype, field->type->size)) == NULL) {
		memset(un, 0, sizeof(*un));
		return (-1);
	}

	/* Decode field */
	if ((flen = (*field->type->decode)(field->type,
	    code + nlen, cmax - nlen, un->un, ebuf, emax)) == -1) {
		FREE(mtype, un->un);
		memset(un, 0, sizeof(*un));
		return (-1);
	}

	/* Set field name */
	*((const char **)&un->field_name) = field->name;
	return (nlen + flen);
}

void
structs_union_free(const struct structs_type *type, void *data)
{
	const char *const mtype = type->args[1].s;
	const struct structs_union *const un = data;
	const struct structs_ufield *const fields = type->args[0].v;
	const struct structs_ufield *field;

	/* Sanity check */
	assert(type->tclass == STRUCTS_TYPE_UNION);

	/* Find field */
	for (field = fields; field->name != NULL
	    && strcmp(un->field_name, field->name) != 0; field++);
	if (field->name == NULL) {
		assert(0);
		return;
	}

	/* Free it */
	(*field->type->uninit)(field->type, un->un);
	FREE(mtype, un->un);
}

